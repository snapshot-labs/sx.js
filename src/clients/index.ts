export { StarkNetTx } from './starknet/starknet-tx';
export { EthereumTx } from './starknet/ethereum-tx';
export { EthereumSig } from './starknet/ethereum-sig';
export { StarkNetSig } from './starknet/starknet-sig';
export { L1Executor } from './starknet/l1-executor';
export { EthereumTx as EvmEthereumTx } from './evm/ethereum-tx';
export { EthereumSig as EvmEthereumSig } from './evm/ethereum-sig';
