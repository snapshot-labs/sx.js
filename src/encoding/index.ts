export * from './slot-key';
export * from './execution-hash';
export * from './calldata';
export * from './starknet-commit';
