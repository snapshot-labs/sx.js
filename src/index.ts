export * as clients from './clients';
export * as utils from './utils';
export { getExecutionData } from './executors';
export * from './strategies';
export * from './networks';
export * from './types';
