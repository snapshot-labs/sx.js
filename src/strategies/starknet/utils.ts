import { keccak256 } from '@ethersproject/keccak256';
import { StaticJsonRpcProvider } from '@ethersproject/providers';
import fetch from 'cross-fetch';

export function getEthRpcUrl(chainId: number) {
  const apiKey = '46a5dd9727bf48d4a132672d3f376146';

  // TODO: ideally we would use rpc.snapshotx.xyz here but those don't support eth_getProof with past lookup
  if (chainId === 1) return `https://mainnet.infura.io/v3/${apiKey}`;
  if (chainId === 5) return `https://goerli.infura.io/v3/${apiKey}`;
  if (chainId === 11155111) return `https://sepolia.infura.io/v3/${apiKey}`;
  if (chainId === 137) return `https://polygon-mainnet.infura.io/v3/${apiKey}`;
  if (chainId === 42161) return `https://arbitrum-mainnet.infura.io/v3/${apiKey}`;

  throw new Error(`Unsupported chainId ${chainId}`);
}

export function getEthProvider(chainId: number) {
  return new StaticJsonRpcProvider(getEthRpcUrl(chainId), chainId);
}

export function getSlotKey(voterAddress: string, slotIndex: number) {
  return keccak256(
    `0x${voterAddress.slice(2).padStart(64, '0')}${slotIndex.toString(16).padStart(64, '0')}`
  );
}

export async function getBinaryTree(
  deployedOnChain: string,
  snapshotTimestamp: number,
  chainId: number
) {
  const res = await fetch(
    `https://ds-indexer.api.herodotus.cloud/binsearch-path?timestamp=${snapshotTimestamp}&deployed_on_chain=${deployedOnChain}&accumulates_chain=${chainId}`,
    {
      headers: {
        accept: 'application/json'
      }
    }
  );

  return res.json();
}
