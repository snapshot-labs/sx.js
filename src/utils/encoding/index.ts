export * from './slot-key';
export * from './execution-hash';
export * from './calldata';
export * from './starknet-commit';
export * from './starknet-execution-params';
export * from './eth-sig';
export * from './storage-vars';
export * from './padding';
