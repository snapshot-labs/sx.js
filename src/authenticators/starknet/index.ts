import createVanillaAuthenticator from './vanilla';
import createEthSigAuthenticator from './ethSig';
import createEthTxAuthenticator from './ethTx';
import createStarkSigAuthenticator from './starkSig';
import createStarkTxAuthenticator from './starkTx';
import { hexPadLeft } from '../../utils/encoding';
import type { Authenticator, NetworkConfig } from '../../types';

export function getAuthenticator(
  address: string,
  networkConfig: NetworkConfig
): Authenticator | null {
  const authenticator = networkConfig.authenticators[hexPadLeft(address)];
  if (!authenticator) return null;

  if (authenticator.type === 'vanilla') {
    return createVanillaAuthenticator();
  }

  if (authenticator.type === 'ethSig') {
    return createEthSigAuthenticator();
  }

  if (authenticator.type === 'ethTx') {
    return createEthTxAuthenticator();
  }

  if (authenticator.type === 'starkSig') {
    return createStarkSigAuthenticator();
  }

  if (authenticator.type === 'starkTx') {
    return createStarkTxAuthenticator();
  }

  return null;
}
