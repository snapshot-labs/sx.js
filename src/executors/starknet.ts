import { createStarknetExecutionParams } from '../utils/encoding';
import type { Call } from 'starknet';

export default function createStarknetExecutor() {
  return {
    type: 'starknet',
    getExecutionData(executorAddress: string, calls: Call[]) {
      return {
        executor: executorAddress,
        executionParams: createStarknetExecutionParams(calls)
      };
    }
  };
}
