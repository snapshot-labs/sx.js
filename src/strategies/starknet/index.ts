import createVanillaStrategy from './vanilla';
import createMerkleWhitelistStrategy from './merkleWhitelist';
import createErc20VotesStrategy from './erc20Votes';
import { hexPadLeft } from '../../utils/encoding';
import type { Strategy, NetworkConfig } from '../../types';

export function getStrategy(address: string, networkConfig: NetworkConfig): Strategy | null {
  const strategy = networkConfig.strategies[hexPadLeft(address)];
  if (!strategy) return null;

  if (strategy.type === 'vanilla') {
    return createVanillaStrategy();
  }

  if (strategy.type === 'whitelist') {
    return createMerkleWhitelistStrategy();
  }

  if (strategy.type === 'erc20Votes') {
    return createErc20VotesStrategy();
  }

  return null;
}
