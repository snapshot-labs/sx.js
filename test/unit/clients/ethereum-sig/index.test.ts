import { defaultProvider } from 'starknet';
import { Wallet } from '@ethersproject/wallet';
import { EthereumSig } from '../../../../src/clients';
import { getStorageVarAddress } from '../../../../src/utils/encoding';
import { Choice } from '../../../../src/utils/choice';

const latestL1Block = 8050780;

describe('EthereumSig', () => {
  expect(process.env.GOERLI_NODE_URL).toBeDefined();

  const ethUrl = process.env.GOERLI_NODE_URL as string;
  const manaUrl = '';
  const starkProvider = defaultProvider;

  const ethSigClient = new EthereumSig({ ethUrl, starkProvider, manaUrl });
  const wallet = new Wallet('be2b70290c687fadeaac651bfc4578948ef25c932f6b0ae6e7f2047ce61bcbaa');
  const walletAddress = wallet.address;
  const space = '0x7e6e9047eb910f84f7e3b86cea7b1d7779c109c970a39b54379c1f4fa395b28';
  const authenticator = '0x64cce9272197eba6353f5bbf060e097e516b411e66e83a9cf5910a08697df14';
  const strategy = 1;
  const executor = '0x4ecc83848a519cc22b0d0ffb70e65ec8dde85d3d13439eff7145d4063cf6b4d';

  beforeAll(() => {
    const getStorageAt = defaultProvider.getStorageAt;
    jest.spyOn(defaultProvider, 'getStorageAt').mockImplementation((contractAddress, key) => {
      if (
        contractAddress === '0x69606dd1655fdbbf8189e88566c54890be8f7e4a3650398ac17f6586a4a336d' &&
        key === getStorageVarAddress('_latest_l1_block')
      ) {
        return Promise.resolve(latestL1Block.toString(16));
      }

      return getStorageAt.call(defaultProvider, contractAddress, key);
    });

    jest.spyOn(ethSigClient, 'generateSalt').mockImplementation(() => 0);
  });

  afterAll(() => {
    jest.restoreAllMocks();
  });

  it('should create propose envelope', async () => {
    const envelope = await ethSigClient.propose(wallet, walletAddress, {
      space,
      authenticator,
      strategies: [strategy],
      executor,
      metadataUri: 'ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca',
      executionParams: []
    });

    expect(envelope).toMatchSnapshot();
  });

  it('should create vote envelope', async () => {
    const envelope = await ethSigClient.vote(wallet, walletAddress, {
      space,
      authenticator,
      strategies: [strategy],
      proposal: 3,
      choice: Choice.FOR
    });

    expect(envelope).toMatchSnapshot();
  });
});
