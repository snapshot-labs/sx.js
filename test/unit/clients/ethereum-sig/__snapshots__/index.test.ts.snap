// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EthereumSig should create propose envelope 1`] = `
{
  "address": "0x0538D033B879aC94C709c1E408CC081345427379",
  "data": {
    "domain": {
      "chainId": 5,
      "name": "snapshot-x",
      "version": "1",
    },
    "message": {
      "authenticator": "0x064cce9272197eba6353f5bbf060e097e516b411e66e83a9cf5910a08697df14",
      "author": "0x0538D033B879aC94C709c1E408CC081345427379",
      "executionHash": "0x049ee3eba8c1600700ee1b87eb599f16716b0b1022947733551fde4050ca6804",
      "executionParams": [],
      "executor": "0x04ecc83848a519cc22b0d0ffb70e65ec8dde85d3d13439eff7145d4063cf6b4d",
      "metadataUri": "ipfs://QmNrm6xKuib1THtWkiN5CKtBEerQCDpUtmgDqiaU2xDmca",
      "salt": 0,
      "space": "0x07e6e9047eb910f84f7e3b86cea7b1d7779c109c970a39b54379c1f4fa395b28",
      "strategies": [
        1,
      ],
      "strategiesHash": "0x078d74f61aeaa8286418fd34b3a12a610445eba11d00ecc82ecac2542d55f7a4",
      "strategiesParamsHash": "0x04c643e1ba37b388f483a9dce226443876069bc87ff7ed6763c1f12048ff5cf9",
    },
    "types": {
      "Propose": [
        {
          "name": "authenticator",
          "type": "bytes32",
        },
        {
          "name": "space",
          "type": "bytes32",
        },
        {
          "name": "author",
          "type": "address",
        },
        {
          "name": "metadata_uri",
          "type": "string",
        },
        {
          "name": "executor",
          "type": "bytes32",
        },
        {
          "name": "execution_hash",
          "type": "bytes32",
        },
        {
          "name": "strategies_hash",
          "type": "bytes32",
        },
        {
          "name": "strategies_params_hash",
          "type": "bytes32",
        },
        {
          "name": "salt",
          "type": "uint256",
        },
      ],
    },
  },
  "sig": "0x33bbb847e30b943aadc6101e20fbd32da3804d2b217767882cacfb85bd5e8d5a1a81dfa00b5ceec8945545276f56dd2a1190ac129b3f39bad5789a4397d8e3621b",
}
`;

exports[`EthereumSig should create vote envelope 1`] = `
{
  "address": "0x0538D033B879aC94C709c1E408CC081345427379",
  "data": {
    "domain": {
      "chainId": 5,
      "name": "snapshot-x",
      "version": "1",
    },
    "message": {
      "authenticator": "0x064cce9272197eba6353f5bbf060e097e516b411e66e83a9cf5910a08697df14",
      "choice": 1,
      "proposal": 3,
      "salt": 0,
      "space": "0x07e6e9047eb910f84f7e3b86cea7b1d7779c109c970a39b54379c1f4fa395b28",
      "strategies": [
        1,
      ],
      "strategiesHash": "0x078d74f61aeaa8286418fd34b3a12a610445eba11d00ecc82ecac2542d55f7a4",
      "strategiesParamsHash": "0x04c643e1ba37b388f483a9dce226443876069bc87ff7ed6763c1f12048ff5cf9",
      "voter": "0x0538D033B879aC94C709c1E408CC081345427379",
    },
    "types": {
      "Vote": [
        {
          "name": "authenticator",
          "type": "bytes32",
        },
        {
          "name": "space",
          "type": "bytes32",
        },
        {
          "name": "voter",
          "type": "address",
        },
        {
          "name": "proposal",
          "type": "uint256",
        },
        {
          "name": "choice",
          "type": "uint256",
        },
        {
          "name": "strategies_hash",
          "type": "bytes32",
        },
        {
          "name": "strategies_params_hash",
          "type": "bytes32",
        },
        {
          "name": "salt",
          "type": "uint256",
        },
      ],
    },
  },
  "sig": "0x529b909b496194a9d4b4fed0f30264943cc25d430352fcf0e5dc0baa7dbeb6f60de20dd6edf8db193bdf39c1c0cfcdf3895927cae31e88c4c28b9eacf736c2401c",
}
`;
